lib/utils/threading.jl

```diff
@@ -37,3 +37,48 @@ macro spinlock(l, ex)
     end
   end
 end
+
+
+"""
+    LazyInitialized{T}() do
+        # initialization, producing a value of type T
+    end
+
+A thread-safe, lazily-initialized wrapper for a value of type `T`. Fetch the value by
+calling `getindex`. The constructor is ensured to only be called once from a single thread.
+
+This type is intended for lazy initialization of e.g. global structures, without using
+`__init__`. It is similar to protecting accesses using a lock, but is much cheaper.
+
+"""
+struct LazyInitialized{T,F}
+    # 0: uninitialized
+    # 1: initializing
+    # 2: initialized
+    guard::Threads.Atomic{Int}
+    value::Base.RefValue{T}
+    constructor::F
+
+    LazyInitialized{T,F}(constructor::F) where {T,F} =
+        new(Threads.Atomic{Int}(0), Ref{T}(), constructor)
+end
+LazyInitialized{T}(constructor::F) where {T,F} = LazyInitialized{T,F}(constructor)
+
+function Base.getindex(x::LazyInitialized)
+    while x.guard[] != 2
+        initialize!(x)
+    end
+    assume(isassigned(x.value)) # to get rid of the check
+    x.value[]
+end
+
+@noinline function initialize!(x::LazyInitialized)
+    status = Threads.atomic_cas!(x.guard, 0, 1)
+    if status == 0
+        x.value[] = x.constructor()
+        x.guard[] = 2
+    else
+        yield()
```

Comment:
I suspect you don't want to use an Event due to the allocation cost? The yield should suffice, but to relies on the scheduler to not always pick the same task or set of same task all live locking.

Reply:
Yeah exactly, it's at 0 allocations now (although I can't seem to get past one remaining 16-byte allocation from somewhere in IdDict...)